src/Features/Core/Portable/Completion/Providers/AbstractPreprocessorCompletionProvider.cs

```diff
@@ -0,0 +1,60 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Completion;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Extensions.ContextQuery;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.Completion.Providers
+{
+    internal abstract class AbstractPreprocessorCompletionProvider : LSPCompletionProvider
+    {
+        protected abstract Task<SyntaxContext> CreateContextAsync(Workspace workspace, SemanticModel semanticModel, int position, System.Threading.CancellationToken cancellationToken);
+
+        public sealed override async Task ProvideCompletionsAsync(CompletionContext context)
+        {
+            var cancellationToken = context.CancellationToken;
+            var originatingDocument = context.Document;
+            var position = context.Position;
+
+            var semanticModel = await originatingDocument.ReuseExistingSpeculativeModelAsync(position, cancellationToken).ConfigureAwait(false);
+
+            var solution = originatingDocument.Project.Solution;
+            var syntaxContext = await CreateContextAsync(solution.Workspace, semanticModel, position, cancellationToken).ConfigureAwait(false);
+            if (!syntaxContext.IsPreProcessorExpressionContext)
+                return;
+
+            using var _ = PooledHashSet<string>.GetInstance(out var preprocessorNames);
+
+            // Walk all the projects this document is linked in so that we get the full set of preprocessor symbols
+            // defined across all of them.
+            foreach (var documentId in solution.GetRelatedDocumentIds(originatingDocument.Id))
+            {
+                var document = solution.GetRequiredDocument(documentId);
+                var currentSyntaxTree = await document.GetRequiredSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+
+                preprocessorNames.AddRange(currentSyntaxTree.Options.PreprocessorSymbolNames);
+            }
+
+            // Keep all the preprocessor symbol names together.  We don't want to intermingle them with any keywords we
+            // include (like `true/false`)
+            var order = 0;
+            foreach (var name in preprocessorNames.OrderBy(a => a))
+            {
+                context.AddItem(CommonCompletionItem.Create(
+                    name,
+                    displayTextSuffix: "",
+                    CompletionItemRules.Default,
+                    glyph: Glyph.Keyword,
+                    sortText: order.ToString("0000")));
```

Comment:
Not sure how I feel about grouping items by assigning random sort text. For one, it make it difficult if another provider want to group its items too. Maybe add a flag to items to be grouped by provider, so when we do comparison we check for ProviderName first?

Reply:
i hear you, but somewhat disagree.  SortText is effectively defined for almsot exactly this purpose.  This sort of "i want to group one group specially" was something we actually intentionally added this field for :)

so it feels fine for me to do this here as we don't have hte need for a more complex solution.